### DISCUSSION TITLE
ECS/Fargate and ServiceDiscovery - multiple container ports

### DISCUSSION THREAD

**Discussion by clarkm5** [aws/aws-cdk#19562] [Q&A]
We're using Fargate + AWS API and CloudMap/ServiceDiscovery

One of our containers exposes two ports - 8080 and 8090.  At the moment, we can only see how to expose one port as a service via Service Discovery (using cloud_map_options as below) - In theory we should be able to add another service configured to hit port 8090.  However, it is not clear how we would create another CloudMap service for the FargateService..  
 
```
        cloud_map_options=ecs.CloudMapOptions(
                container_port= 8080,
                cloud_map_namespace=dns_namespace,
                dns_record_type=servicediscovery.DnsRecordType.SRV
            )
        )
```

Has anyone had any success with this?

**Comment by indrora** [comment]
This construct only supports one port/service per container.

**Reply by clarkm5** [reply]
well worth the wait for that answer

**Reply by peterwoodworth** [reply]
CloudFormation (and the ECS `createService` operation) take in an [array](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecs-service.html#cfn-ecs-service-serviceregistries) of `serviceRegistries`. Does it work to specify multiple ports if you have multiple entries in this array that point to the same Service Registry? You could use escape hatches within an aspect to modify the Service definition in the template such that it contains another serviceRegistry entry. If this works, we could make this a feature request to make this possible without escape hatches

**Comment by xlc** [comment]
There is [associateCloudMapService](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ecs.Ec2Service.html#associatewbrcloudwbrmapwbrserviceoptions) which looks like a way to add new container & port. I will give it a try and see if it works.

**Reply by xlc** [reply]
Well, it doesn't work. https://github.com/aws/aws-cdk/blob/f3970718ff8b4571bcfef6ebc0f480cac14e47ee/packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts#L1340 prevents it and I don't really understand why the code is written in such way.

**Reply by xlc** [reply]
I guess this is just impossible

<img width="819" alt="Screenshot 2023-10-29 at 7 27 02 PM" src="https://github.com/aws/aws-cdk/assets/1201310/a4803fad-4973-4c98-a520-a0cd21a46b6f">
